### PR TITLE
[Bug Fix] Hunyuan image cfg bugfix

### PR DESCRIPTION
**Pull request by AbelSara**
## Purpose
Fix a correctness bugs
 in CFG-parallel KV transfer that broke Hunyuan-Image (and other CFG-parallel diffusion models where the AR stage does not split its KV cache by CFG branch):                                                                            
1. Silent wrong output on CFG follower rank. When the AR/prior stage produces a single shared KV cache (no per-branch   cfg_branch_roles), _build_cfg_rank_local_payloads padded every rank beyond rank 0 with None. With cfg_size == 2, the CFG follower (rank 1) therefore received no KV cache and could not reuse the shared positive prefix.                                                                                                      
        
## Design
- kv_transfer_manager.py — follower padding fix: In `_build_cfg_rank_local_payloads`, when there are no branch roles, cfg_size == 2, and a non-empty main_payload exists, pad the follower rank with a copy of the main (positive) payload instead of None. This  gives the CFG follower the same shared-prefix KV it needs for reuse. The fix is deliberately narrow:                            
    - cfg_size > 2 with no branch roles still pads with None (no safe single source to fan out).                                  
    - An empty main_payload still pads with None.                                                                                 
    - The existing branch-role path (where each rank gets its own branch KV) is untouched.                                        
- diffusion_model_runner.py — fail fast: Capture the kv_received return value from both the prefetch and sync receive paths, and raise `RuntimeError `when `config.need_recv_cache` is set but nothing was received, so a misconfigured / failed transfer fails loudly instead of producing garbage.   
## Test Plan
observe neg kv reuse

### Log
image

### Performace
- before
image

- after 
image

- speed up: +1.5%

**BEFORE SUBMITTING:** read [CONTRIBUTING.md](https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md) and run the [precheck-pr skill](https://github.com/vllm-project/vllm-omni/blob/main/.claude/skills/precheck-pr/SKILL.md) with the code agent for a self-check against project conventions.
(anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by AbelSara** [comment]
This PR is ready for review @Gaohan123 @hsliuustc0106 @Bounty-hunter

**Comment by Bounty-hunter** [comment]
For the First bug, I remember `distribute_kv_cache` will broadcase kv from cfg/sp rank 0 to other rank. Was this issue introduced by another PR?

**Comment by AbelSara** [comment]
> For the First bug, I remember `distribute_kv_cache` will broadcase kv from cfg/sp rank 0 to other rank. Was this issue introduced by another PR?

It’s not a newly introduced bug. In the CFG scenario, data has always been sent out, but it’s None.

File: vllm_omni/diffusion/distributed/group_coordinator.py [hunk 1]
```diff
@@ -589,11 +589,12 @@ def send(self, tensor: torch.Tensor, dst: int | None = None) -> None:
         if dst is None:
             dst = self.group_next_rank
 
-        torch.distributed.send(
-            tensor,
-            self.ranks[dst],
-            group=(self.device_groups[self.rank_in_group % 2] if self.world_size == 2 else self.device_group),
-        )
```
Comment (Bounty-hunter) [inline vllm_omni/diffusion/distributed/group_coordinator.py:596]: This change is functionally equivalent for send/recv , but it fixes a AttributeError, since self.device_groups is not defined in `GroupCoordinator`. so it not a silent error?  It feels a bit odd for the base class contain the logic for a subclass.

**Review by Bounty-hunter — APPROVED** [review]
LGTM

File: vllm_omni/diffusion/distributed/group_coordinator.py (hunk 1 above)
Comment (AbelSara) [reply vllm_omni/diffusion/distributed/group_coordinator.py:596]: Thanks for your review.  The base class shouldn't use `getattr` to probe subclass attributes. In a follow-up commit, the device_groups selection logic has been moved into the overridden `PipelineGroupCoordinator.send/recv`; the base class now directly uses `self.device_group`. The behavior of all paths remains unchanged.

**Review by wtomin — APPROVED** [review]
LGTM.

**Review by Gaohan123 — COMMENTED** [review]
In PR description, you claim that there is a cache check mechanism in modelrunner, but I don't find it in your code changes. Please double check

**Comment by AbelSara** [comment]
> In PR description, you claim that there is a cache check mechanism in modelrunner, but I don't find it in your code changes. Please double check

I updated the description because I think that when a worker fails to receive, the process should still be completed, rather than having a single worker throw an exception and cause other workers to deadlock.

**Review by lishunyang12 — COMMENTED** [review]
mostly small, but the shallow-copied `past_key_values` shared across ranks is the one worth a look.

**Review by lishunyang12 — COMMENTED** [review]
Fix for the cfg KV reuse looks right. Two things on the padding copy and the new fallback log, plus the new test only covers the happy path.

File: vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py [hunk 2]
```diff
@@ -2919,6 +2925,10 @@ def _maybe_handle_ar_kv_reuse(
     ):
         ar_kv_data = model_kwargs.pop("ar_kv_data", None)
         if ar_kv_data is None:
+            logger.info(
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py]: This should be `logger.debug` like the neg-branch log above — it can fire every denoise step whenever AR KV is absent.

File: vllm_omni/distributed/omni_connectors/kv_transfer_manager.py [hunk 3]
```diff
@@ -856,8 +856,15 @@ def _build_cfg_rank_local_payloads(cls, req: Any, cfg_size: int) -> list[dict[st
 
             payloads.append(local_payload)
 
+        padding = None
+        if not branch_roles and cfg_size == 2 and main_payload:
+            # AR didn't split KV by branch; cfg follower needs the same
+            # positive KV for shared-prefix reuse. Only safe for cfg_size == 2:
+            # for cfg_size > 2 every rank must get its own KV, which we can't
+            # assign here, so padding stays None and the runner raises.
+            padding = dict(main_payload)
```
Comment (lishunyang12) [inline vllm_omni/distributed/omni_connectors/kv_transfer_manager.py:870]: `dict(main_payload)` is a shallow copy, so the follower shares rank 0's `past_key_values` tensors. In-place mutation on one branch clobbers the other. Deep-copy the tensors.

File: tests/diffusion/test_diffusion_model_runner.py [hunk 4]
```diff
@@ -509,6 +509,25 @@ def test_split_diffusion_output_by_request_slices_tuple_outputs():
     assert torch.equal(outputs[1].output[1], torch.tensor([20]))
 
 
+@pytest.mark.core_model
+@pytest.mark.cpu
+def test_execute_model_runs_forward_after_kv_receive(monkeypatch):
```
Comment (lishunyang12) [inline tests/diffusion/test_diffusion_model_runner.py:514]: This only covers the happy path. Add a case exercising the fail-fast `RuntimeError` when the KV receive yields no payload.

File: vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py (hunk 2 above)
Comment (AbelSara) [reply vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py]: fixed

File: vllm_omni/distributed/omni_connectors/kv_transfer_manager.py (hunk 3 above)
Comment (AbelSara) [reply vllm_omni/distributed/omni_connectors/kv_transfer_manager.py:870]: deepcopy is costly, so we use a comment to call attention to it.

File: tests/diffusion/test_diffusion_model_runner.py (hunk 4 above)
Comment (AbelSara) [reply tests/diffusion/test_diffusion_model_runner.py:514]: this pr does not contain `RuntimeError` raise
